File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    static List<Integer>[] graph = new ArrayList[102];
    static boolean[][] cases = new boolean[102][102];
    static boolean[][] cutted = new boolean[102][102];
    
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;
        
        
        for(int i=0;i<=n;i++){
            graph[i] = new ArrayList<Integer>();
        }
        
        for(int i=0;i<wires.length;i++){
            int[] wire = wires[i];
            graph[wire[0]].add((Integer)wire[1]);
            graph[wire[1]].add((Integer)wire[0]);
        }
        
        for(int i=1;i<=n;i++){
            for(int j=0;j<graph[i].size();j++){
                int num = graph[i].get(j);
                if(cases[i][num]) continue;
                cases[i][num] = true;
                cases[num][i] = true;
                
                cutted[i][num] = true;
                cutted[num][i] = true;
                
                int cnt = getCnt(i);
                int other = n-cnt;
                
                System.out.printf("i: %d j: %d cnt: %d\n", i,num,cnt);
                
                answer = Math.min(answer, Math.abs(cnt - other));
                
                cutted[i][num] = false;
                cutted[num][i] = false;
            }
        }
        
        return answer;
    }
    
    static int getCnt(int start){
        boolean[] vis = new boolean[102];
        LinkedList<Integer> q = new LinkedList<>();
        int cnt = 0;
        
        q.offer((Integer)start);
        vis[start] = true;
        cnt++;
        
        while(!q.isEmpty()){
            int cur = (int)q.poll();
            
            int n = graph[cur].size();
            
            for(int i=0;i<n;i++){
                int num = (int)graph[cur].get(i);
                
                if(vis[num]) continue;
                if(cutted[cur][num]) continue;
                q.offer((Integer)num);
                cnt++;
                vis[num] = true;
            }
        }
        
        
        return cnt;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.kt
import java.util.*
import kotlin.math.*

val q = LinkedList<Int>()

val removed = Array(101){Array(101){false}}

val graph = Array(101){HashSet<Int>()}

class Solution {
    fun solution(n: Int, wires: Array<IntArray>): Int {
        var answer: Int = Int.MAX_VALUE
        
        wires.forEach{
            val v1 = it[0]
            val v2 = it[1]
            
            graph[v1].add(v2)
            graph[v2].add(v1)
        }
        
        for(i in 1..n){
            graph[i].forEach{
                removed[i][it] = true
                removed[it][i] = true
                val cnt1 = bfs(i)
                val cnt2 = bfs(it)
                answer = minOf(answer, abs(cnt1-cnt2))
                removed[i][it] = false
                removed[it][i] = false
            }
        }
        
        return answer
    }
    
    fun bfs(start:Int):Int{
        val vis = Array(101){false}
        q.offer(start)
        vis[start] = true
        var cnt = 1
        
        while(!q.isEmpty()){
            val cur = q.poll()
            graph[cur].forEach{
                if(removed[cur][it]) return@forEach
                if(vis[it]) return@forEach
                cnt++
                q.offer(it)
                vis[it] = true
            }
        }
        return cnt
    }
    
    
    
    
}